pkg/utils/kubernetes/reduce/filters.go

```diff
@@ -405,3 +406,49 @@ func filterKongPluginBindings(kpbs []configurationv1alpha1.KongPluginBinding) []
 
 	return append(kpbs[:best], kpbs[best+1:]...)
 }
+
+// -----------------------------------------------------------------------------
+// Filter functions - KongCredentials
+// -----------------------------------------------------------------------------
+
+// filterKongCredentials filters out the KongCredentials to be kept and returns all the KongCredentials
+// to be deleted.
+// The KongPluginBinding with Programmed status condition is kept.
+// If no such binding is found the oldest is kept.
+func filterKongCredentials[
+	T constraints.SupportedCredentialType,
+	TPtr constraints.KongCredential[T],
+](creds []T) []T {
+	if len(creds) < 2 {
+		return []T{}
+	}
+
+	programmed := -1
+	best := 0
+	for i, cred := range creds {
+		ptr := TPtr(&cred)
+		if lo.ContainsBy(ptr.GetConditions(), func(c metav1.Condition) bool {
+			return c.Type == konnectv1alpha1.KonnectEntityProgrammedConditionType &&
+				c.Status == metav1.ConditionTrue
+		}) {
+
+			ptrProgrammed := TPtr(&creds[programmed])
```

Comment:
Is it possible to run into `Out of Bound` panic as the `programmed` is initialized to `-1` and the indexing happens before changing the value of `programmed`?

Reply:
Good call, I added a test for that and fixed the implementation in 8210664aae17ca191a3d0c2455cb5136c0d65037. PTAL